src/Microsoft.Extensions.DependencyInjection/ServiceLookup/CallSiteValidator.cs

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.Extensions.DependencyInjection.ServiceLookup
+{
+    internal class CallSiteValidator: CallSiteVisitor<CallSiteValidator.CallSiteValidatorState, Type>
+    {
+        private readonly Dictionary<Type, Type> _scopedServices = new Dictionary<Type, Type>();
+
+        public void ValidateCallSite(Type serviceType, IServiceCallSite callSite)
+        {
+            var scoped = VisitCallSite(callSite, default(CallSiteValidatorState));
+            if (scoped != null)
+            {
+                _scopedServices.Add(serviceType, scoped);
+            }
+        }
+
+        public void ValidateResolution(Type serviceType, ServiceProvider serviceProvider)
+        {
+            Type scopedService;
+            if (ReferenceEquals(serviceProvider, serviceProvider.Root)
+                && _scopedServices.TryGetValue(serviceType, out scopedService))
+            {
+                throw new InvalidOperationException(Resources.FormatScopedResolvedFromRootException(
+                    serviceType,
+                    scopedService,
```

Comment:
Can't `serviceType` and `scopedService` be the same? That would make the message unnecessarily confusing.